### PR TITLE
Enable stream in HIP

### PR DESCRIPTION
**Pull request by Rombur**
Most of the code for streams in HIP is already in develop, we just need to allow non-default streams.

### CODE REVIEW

**Comment by masterleinad** [comment]
What about a test for this? Can you modify `core/unit_test/cuda/TestCuda_InterOp_Streams.cpp` accordingly?

File: core/src/HIP/Kokkos_HIP_Space.cpp [hunk 1]
```diff
@@ -92,25 +98,25 @@ DeepCopy<Kokkos::Experimental::HIPSpace, HostSpace,
...
 DeepCopy<Kokkos::Experimental::HIPSpace, Kokkos::Experimental::HIPSpace,
          Kokkos::Experimental::HIP>::DeepCopy(const Kokkos::Experimental::HIP&
-                                              /*instance*/,
+                                                  instance,
                                               void* dst, const void* src,
                                               size_t n) {
-  // FIXME_HIP use instance
-  HIP_SAFE_CALL(hipMemcpy(dst, src, n, hipMemcpyDefault));
+  instance.fence();
+  DeepCopyAsyncHIP(dst, src, n);
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Space.cpp:105]: You've got to use the stream provided by the instance, not the one from `get_deep_copy_stream()`

File: core/src/Kokkos_HIP_Space.hpp [hunk 2]
```diff
@@ -657,7 +657,7 @@ class HIP {
 
   ~HIP() = default;
   HIP();
-  //  explicit HIP( const int instance_id );
+  HIP(hipStream_t stream);
```
[the remark this answers is not in the document]
Comment (Rombur) [reply core/src/Kokkos_HIP_Space.hpp:660]: I didn't drop anything this is a new constructor that we did not have before. Since the new constructor uses an `hipStream_t` instead of an `int` I don't see the point of using `explicit`

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_HIP_Space.hpp:660]: Overlook on my part.  Sorry :)

File: core/src/HIP/Kokkos_HIP_Space.cpp (hunk 1 above)
Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_Space.cpp:105]: That's not how it works in CUDA. I can change it but then the behavior will be different between CUDA and HIP. No matter what's your instance, we always use the same stream for the `deep_copy`.

Comment (dalg24) [reply core/src/HIP/Kokkos_HIP_Space.cpp:105]: Are you sure?
https://github.com/kokkos/kokkos/blob/b973b126183f6ea756c7b02591cecaa445e85a8f/core/src/Cuda/Kokkos_CudaSpace.cpp#L99-L103

What am I missing?

Comment (Rombur) [reply core/src/HIP/Kokkos_HIP_Space.cpp:105]: Hmm I see. I used https://github.com/kokkos/kokkos/blob/develop/core/src/Kokkos_CudaSpace.hpp#L547-L551

**Comment by Rombur** [comment]
I have added a test and done the other changes. The current code does leak as does the CUDA code. I need to think of a clean way to fix that for both backend.

**Comment by dalg24** [comment]
> I have added a test and done the other changes. The current code does leak as does the CUDA code. I need to think of a clean way to fix that for both backend.

Would you please open an issue and add it to the agenda for tomorrow's developer meeting?

**Comment by dalg24** [comment]
Pasting the diff here for reference
```diff
--- core/unit_test/hip/TestHIP_InterOp_Streams.cpp
+++ core/unit_test/cuda/TestCuda_InterOp_Streams.cpp
@@ -43,7 +43,7 @@
 */

 #include <Kokkos_Core.hpp>
-#include <hip/TestHIP_Category.hpp>
+#include <cuda/TestCuda_Category.hpp>

 namespace Test {

@@ -56,10 +56,9 @@

 namespace {
 struct FunctorRange {
-  Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
-               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+  Kokkos::View<int*, Kokkos::CudaSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>
       a;
-  FunctorRange(Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
+  FunctorRange(Kokkos::View<int*, Kokkos::CudaSpace,
                             Kokkos::MemoryTraits<Kokkos::Unmanaged>>
                    a_)
       : a(a_) {}
@@ -68,10 +67,9 @@
   void operator()(const int i) const { a(i) += 1; }
 };
 struct FunctorRangeReduce {
-  Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
-               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+  Kokkos::View<int*, Kokkos::CudaSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>
       a;
-  FunctorRangeReduce(Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
+  FunctorRangeReduce(Kokkos::View<int*, Kokkos::CudaSpace,
                                   Kokkos::MemoryTraits<Kokkos::Unmanaged>>
                          a_)
       : a(a_) {}
@@ -80,10 +78,9 @@
   void operator()(const int i, int& lsum) const { lsum += a(i); }
 };
 struct FunctorMDRange {
-  Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
-               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+  Kokkos::View<int*, Kokkos::CudaSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>
       a;
-  FunctorMDRange(Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
+  FunctorMDRange(Kokkos::View<int*, Kokkos::CudaSpace,
                               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
                      a_)
       : a(a_) {}
@@ -92,10 +89,9 @@
   void operator()(const int i, const int j) const { a(i * 10 + j) += 1; }
 };
 struct FunctorMDRangeReduce {
-  Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
-               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+  Kokkos::View<int*, Kokkos::CudaSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>
       a;
-  FunctorMDRangeReduce(Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
+  FunctorMDRangeReduce(Kokkos::View<int*, Kokkos::CudaSpace,
                                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>
                            a_)
       : a(a_) {}
@@ -106,18 +102,16 @@
   }
 };
 struct FunctorTeam {
-  Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
-               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+  Kokkos::View<int*, Kokkos::CudaSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>
       a;
-  FunctorTeam(Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
+  FunctorTeam(Kokkos::View<int*, Kokkos::CudaSpace,
                            Kokkos::MemoryTraits<Kokkos::Unmanaged>>
                   a_)
       : a(a_) {}

   KOKKOS_INLINE_FUNCTION
   void operator()(
-      const Kokkos::TeamPolicy<Kokkos::Experimental::HIP>::member_type& team)
-      const {
+      const Kokkos::TeamPolicy<Kokkos::Cuda>::member_type& team) const {
     int i = team.league_rank();
     Kokkos::parallel_for(Kokkos::TeamThreadRange(team, 10),
                          [&](const int j) { a(i * 10 + j) += 1; });
@@ -125,18 +119,16 @@
 };

 struct FunctorTeamReduce {
-  Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
-               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+  Kokkos::View<int*, Kokkos::CudaSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>
       a;
-  FunctorTeamReduce(Kokkos::View<int*, Kokkos::Experimental::HIPSpace,
+  FunctorTeamReduce(Kokkos::View<int*, Kokkos::CudaSpace,
                                  Kokkos::MemoryTraits<Kokkos::Unmanaged>>
                         a_)
       : a(a_) {}

   KOKKOS_INLINE_FUNCTION
-  void operator()(
-      const Kokkos::TeamPolicy<Kokkos::Experimental::HIP>::member_type& team,
-      int& lsum) const {
+  void operator()(const Kokkos::TeamPolicy<Kokkos::Cuda>::member_type& team,
+                  int& lsum) const {
     int i = team.league_rank();
     int team_sum;
     Kokkos::parallel_reduce(
@@ -147,64 +139,63 @@
 };
 }  // namespace

-// Test Interoperability with HIP Streams
-TEST(hip, raw_hip_streams) {
-  hipStream_t stream;
-  hipStreamCreate(&stream);
+// Test Interoperability with Cuda Streams
+TEST(cuda, raw_cuda_streams) {
+  cudaStream_t stream;
+  cudaStreamCreate(&stream);
   Kokkos::InitArguments arguments{-1, -1, -1, false};
   Kokkos::initialize(arguments);
   int* p;
-  hipMalloc(&p, sizeof(int) * 100);
+  cudaMalloc(&p, sizeof(int) * 100);

   {
-    Kokkos::Experimental::HIP hip0(stream);
-    Kokkos::View<int*, Kokkos::Experimental::HIPSpace> v(p, 100);
-    Kokkos::deep_copy(hip0, v, 5);
+    Kokkos::Cuda cuda0(stream);
+    Kokkos::View<int*, Kokkos::CudaSpace> v(p, 100);
+    Kokkos::deep_copy(cuda0, v, 5);
     int sum;

-    Kokkos::parallel_for(
-        "Test::hip::raw_hip_stream::Range",
-        Kokkos::RangePolicy<Kokkos::Experimental::HIP>(hip0, 0, 100),
-        FunctorRange(v));
+    Kokkos::parallel_for("Test::cuda::raw_cuda_stream::Range",
+                         Kokkos::RangePolicy<Kokkos::Cuda>(cuda0, 0, 100),
+                         FunctorRange(v));
     Kokkos::parallel_reduce(
-        "Test::hip::raw_hip_stream::RangeReduce",
-        Kokkos::RangePolicy<Kokkos::Experimental::HIP>(hip0, 0, 100),
+        "Test::cuda::raw_cuda_stream::RangeReduce",
+        Kokkos::RangePolicy<Kokkos::Cuda, Kokkos::LaunchBounds<128, 2>>(cuda0,
+                                                                        0, 100),
         FunctorRangeReduce(v), sum);
-    hip0.fence();
+    cuda0.fence();
     ASSERT_EQ(600, sum);

-    Kokkos::parallel_for(
-        "Test::hip::raw_hip_stream::MDRange",
-        Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2>>(
-            hip0, {0, 0}, {10, 10}),
-        FunctorMDRange(v));
-    Kokkos::parallel_reduce(
-        "Test::hip::raw_hip_stream::MDRangeReduce",
-        Kokkos::MDRangePolicy<Kokkos::Experimental::HIP, Kokkos::Rank<2>>(
-            hip0, {0, 0}, {10, 10}),
-        FunctorMDRangeReduce(v), sum);
-    hip0.fence();
+    Kokkos::parallel_for("Test::cuda::raw_cuda_stream::MDRange",
+                         Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2>>(
+                             cuda0, {0, 0}, {10, 10}),
+                         FunctorMDRange(v));
+    Kokkos::parallel_reduce("Test::cuda::raw_cuda_stream::MDRangeReduce",
+                            Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2>,
+                                                  Kokkos::LaunchBounds<128, 2>>(
+                                cuda0, {0, 0}, {10, 10}),
+                            FunctorMDRangeReduce(v), sum);
+    cuda0.fence();
     ASSERT_EQ(700, sum);

-    Kokkos::parallel_for(
-        "Test::hip::raw_hip_stream::Team",
-        Kokkos::TeamPolicy<Kokkos::Experimental::HIP>(hip0, 10, 10),
-        FunctorTeam(v));
+    Kokkos::parallel_for("Test::cuda::raw_cuda_stream::Team",
+                         Kokkos::TeamPolicy<Kokkos::Cuda>(cuda0, 10, 10),
+                         FunctorTeam(v));
     Kokkos::parallel_reduce(
-        "Test::hip::raw_hip_stream::Team",
-        Kokkos::TeamPolicy<Kokkos::Experimental::HIP>(hip0, 10, 10),
+        "Test::cuda::raw_cuda_stream::Team",
+        Kokkos::TeamPolicy<Kokkos::Cuda, Kokkos::LaunchBounds<128, 2>>(cuda0,
+                                                                       10, 10),
         FunctorTeamReduce(v), sum);
-    hip0.fence();
+    cuda0.fence();
     ASSERT_EQ(800, sum);
   }
   Kokkos::finalize();
   offset_streams<<<100, 64, 0, stream>>>(p);
-  HIP_SAFE_CALL(hipDeviceSynchronize());
-  hipStreamDestroy(stream);
+  CUDA_SAFE_CALL(cudaDeviceSynchronize());
+  cudaStreamDestroy(stream);

   int* h_p = new int[100];
-  hipMemcpy(h_p, p, sizeof(int) * 100, hipMemcpyDefault);
-  HIP_SAFE_CALL(hipDeviceSynchronize());
+  cudaMemcpy(h_p, p, sizeof(int) * 100, cudaMemcpyDefault);
+  CUDA_SAFE_CALL(cudaDeviceSynchronize());
   int64_t sum        = 0;
   int64_t sum_expect = 0;
   for (int i = 0; i < 100; i++) {

File: core/src/HIP/Kokkos_HIP_Space.cpp [hunk 3]
```diff
@@ -694,15 +694,21 @@ HIP::HIP() : m_space_instance(&Impl::HIPInternal::singleton()) {
...
 }
 
-// HIP::HIP( const int instance_id )
-//  : m_device( Impl::HIPInternal::singleton().m_hipDev )
-//{}
+HIP::HIP(hipStream_t const stream) : m_space_instance(new Impl::HIPInternal) {
+  Impl::HIPInternal::singleton().verify_is_initialized(
+      "HIP instance constructor");
+  m_space_instance->initialize(Impl::HIPInternal::singleton().m_hipDev, stream);
+}
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/HIP/Kokkos_HIP_Space.cpp:701]: OK with resolving this elsewhere

**Comment by masterleinad** [comment]
> I need to think of a clean way to fix that for both backend.

Can't you just store a bool saving whether the `HIP` resp. `Cuda` class was initialized with an explicit stream or not and clean up in the destructor if necessary?

**Comment by masterleinad** [comment]
Retest this please.

**Comment by Rombur** [comment]
> Can we add the memory leak fix merged for CUDA to this?

Yes, of course.

**Comment by Rombur** [comment]
This is ready for review.

File: core/unit_test/cuda/TestCuda_InterOp_Streams.cpp [hunk 4]
```diff
@@ -156,35 +63,35 @@ TEST(cuda, raw_cuda_streams) {
 
     Kokkos::parallel_for("Test::cuda::raw_cuda_stream::Range",
                          Kokkos::RangePolicy<Kokkos::Cuda>(cuda0, 0, 100),
-                         FunctorRange(v));
+                         FunctorRange<Kokkos::CudaSpace>(v));
```
Comment (dalg24) [inline core/unit_test/cuda/TestCuda_InterOp_Streams.cpp:66]: Any reason to prefer this to deducing execution and memory spaces from the TEST_EXECSPACE macro?

Comment (Rombur) [reply core/unit_test/cuda/TestCuda_InterOp_Streams.cpp:66]: No. Does it matter?

Comment (dalg24) [reply core/unit_test/cuda/TestCuda_InterOp_Streams.cpp:66]: To me yes.

Comment (Rombur) [reply core/unit_test/cuda/TestCuda_InterOp_Streams.cpp:66]: Why though?

Comment (dalg24) [reply core/unit_test/cuda/TestCuda_InterOp_Streams.cpp:66]: Because it hurts maintainability.
If a brave soul decide to work on generating the test .cpp files from CMake he will have to diff these guys to understand what is going on.  I am open to defer doing that if you add a comment as a  weak attempt to prevent someone else to edit one file but not the other.
I do see that it is not as simple as the definition of these two spaces because of things like https://github.com/kokkos/kokkos/blob/f376c99423f4f794109d45b1d96bdd8abd91272d/core/unit_test/cuda/TestCuda_InterOp_Streams.cpp#L100-L105 and I understand why you would be rightfully reluctant to deal with it.

TLDR add a comment where you think is appropriate and you've got my approval.

Comment (Rombur) [reply core/unit_test/cuda/TestCuda_InterOp_Streams.cpp:66]: > I understand why you are reluctant to deal with it

I am not reluctant. I just wanted to know the reason why you wanted this change.
